In [2]:
from tensorflow import keras
import pandas as pd
from sklearn.metrics import mean_squared_error # 우리 모델 훈련기준은 MSE 
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

import warnings
warnings.filterwarnings('ignore')

In [24]:
df = pd.read_csv(r'C:\Users\ITSC\Desktop\project\07.ML\data/boston.csv')

In [25]:
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  PRICE  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  PRICE    506 non-null    float64
dtypes: float64(14)
memory usage: 55.5 KB


- null값 없음

In [27]:
df.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT       PRICE  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

# 데이터 분할

In [28]:
x = df.drop('PRICE', axis = 1)
y = df['PRICE']

In [29]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state= 12)

In [30]:
x_sub, x_val, y_sub, y_val = train_test_split(x_train, y_train, test_size= 0.2, random_state=12)

In [31]:
# 훈련 데이터
x_sub.shape

(323, 13)

# 모델 설계
- 원래는 데이터 분석 들어가야 함

### 확률이 필요할 때
- softmax, sigmoid 

### 지금 우리는 확률 말고 가격이 필요함 (값이 하나만 필요)
- linear

In [32]:
model = keras.Sequential()

In [33]:
# 은닉층1
model.add(keras.layers.Dense(30, activation = 'relu', input_shape = (13, )))

In [34]:
# 은닉층2
model.add(keras.layers.Dense(6, activation = 'relu'))

In [35]:
# 출력층 
model.add(keras.layers.Dense(1))

In [36]:
early_stopping_cb = keras.callbacks.EarlyStopping(patience = 16, restore_best_weights = True)

In [37]:
model.compile(optimizer= 'adam', loss = 'mean_squared_error', metrics= ['mae'])

- loss 모를때는 keras 홈페이지 - api3 API - losses

In [38]:
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_5 (Dense)                      │ (None, 30)                  │             420 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_6 (Dense)                      │ (None, 6)                   │             186 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_7 (Dense)                      │ (None, 1)                   │               7 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 613 (2.39 KB)

 Trainable params: 613 (2.39 KB)

 Non-trainable params: 0 (0.00 B)

In [39]:
history = model.fit(x_sub, y_sub, epochs = 200, validation_data=(x_val, y_val),
                   callbacks= [early_stopping_cb])

Epoch 1/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 1s 16ms/step - loss: 1053.8904 - mae: 29.8299 - val_loss: 146.2342 - val_mae: 10.5431
Epoch 2/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 163.5134 - mae: 10.4516 - val_loss: 172.9665 - val_mae: 10.4260
Epoch 3/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 163.8441 - mae: 9.9656 - val_loss: 116.5463 - val_mae: 8.3380
Epoch 4/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 126.1230 - mae: 9.0400 - val_loss: 92.3910 - val_mae: 7.9854
Epoch 5/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 119.0646 - mae: 8.9286 - val_loss: 87.1733 - val_mae: 7.6493
Epoch 6/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 105.2352 - mae: 8.1247 - val_loss: 85.6819 - val_mae: 7.1643
Epoch 7/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 90.9532 - mae: 7.5260 - val_loss: 78.6482 - val_mae: 6.8325
Epoch 8/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 86.5631 - mae: 7.2543 - val_loss: 72.5955 - val_mae: 6.6311
Epoch 9/200
11/11 ━━━━━━━━━━━━━━━

11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 49.8727 - mae: 4.7858 - val_loss: 40.1641 - val_mae: 4.7076
Epoch 55/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 47.2969 - mae: 4.6895 - val_loss: 39.3521 - val_mae: 4.7869
Epoch 56/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 42.7049 - mae: 4.8901 - val_loss: 42.7996 - val_mae: 4.7862
Epoch 57/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 45.9663 - mae: 4.4427 - val_loss: 40.8779 - val_mae: 4.8949
Epoch 58/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 42.6618 - mae: 4.9490 - val_loss: 38.8014 - val_mae: 4.5258
Epoch 59/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 39.1291 - mae: 4.3754 - val_loss: 37.2397 - val_mae: 4.5057
Epoch 60/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 33.9046 - mae: 4.0853 - val_loss: 37.2158 - val_mae: 4.6853
Epoch 61/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 42.2741 - mae: 4.6497 - val_loss: 39.4698 - val_mae: 4.5524
Epoch 62/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step -

11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 29.0867 - mae: 4.1056 - val_loss: 29.8149 - val_mae: 4.1961
Epoch 108/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 35.7785 - mae: 4.3662 - val_loss: 28.9353 - val_mae: 4.1050
Epoch 109/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 40.5770 - mae: 5.1018 - val_loss: 42.3059 - val_mae: 4.7151
Epoch 110/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 31.0782 - mae: 4.0053 - val_loss: 37.5194 - val_mae: 5.0116
Epoch 111/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 39.8176 - mae: 5.0469 - val_loss: 33.7033 - val_mae: 4.2318
Epoch 112/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 33.2017 - mae: 4.1116 - val_loss: 30.1044 - val_mae: 4.1422
Epoch 113/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 31.4622 - mae: 3.7956 - val_loss: 29.4095 - val_mae: 4.1911
Epoch 114/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 26.7093 - mae: 3.9858 - val_loss: 29.9900 - val_mae: 4.0546
Epoch 115/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 5m

Epoch 160/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 22.8319 - mae: 3.5382 - val_loss: 26.2104 - val_mae: 3.9808
Epoch 161/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 26.5508 - mae: 3.7921 - val_loss: 24.3826 - val_mae: 3.8207
Epoch 162/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 27.5660 - mae: 4.0341 - val_loss: 24.7626 - val_mae: 3.8736
Epoch 163/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 24.6216 - mae: 3.9003 - val_loss: 25.8646 - val_mae: 3.8395
Epoch 164/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 28.3422 - mae: 3.9660 - val_loss: 25.6513 - val_mae: 3.8154
Epoch 165/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 24.3141 - mae: 3.6094 - val_loss: 25.6601 - val_mae: 3.9291
Epoch 166/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 27.4536 - mae: 3.8078 - val_loss: 28.1654 - val_mae: 4.2041
Epoch 167/200
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 28.7624 - mae: 4.4721 - val_loss: 29.3600 - val_mae: 4.0770
Epoch 168/200
11/11 ━━━━━━━━━━━━

# 모델 성능 평가

In [40]:
model.evaluate(x_test, y_test)

4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 27.5187 - mae: 4.0190 


[28.296663284301758, 4.097434043884277]

In [41]:
# 첫번째 예측 값 25.25
model.predict(x_test.iloc[[0]])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step


array([[25.256056]], dtype=float32)

In [42]:
# 실제 정답 값
# 꽤 그럴싸하게 잘맞춤 
y_test.iloc[[0]]

334    20.7
Name: PRICE, dtype: float64

In [43]:
y_pred = model.predict(x_test).flatten()

4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 1000us/step


In [44]:
for i in range(10):
    label = y_test.iloc[i]
    prediction = y_pred[i]
    print(f'실제가격 : {label:.3f}, 예상가격 : {prediction:.3f}')

실제가격 : 20.700, 예상가격 : 25.256
실제가격 : 12.700, 예상가격 : 18.997
실제가격 : 8.500, 예상가격 : 6.259
실제가격 : 25.100, 예상가격 : 27.552
실제가격 : 28.200, 예상가격 : 36.618
실제가격 : 22.500, 예상가격 : 17.621
실제가격 : 18.200, 예상가격 : 23.064
실제가격 : 43.500, 예상가격 : 34.083
실제가격 : 36.100, 예상가격 : 31.991
실제가격 : 23.800, 예상가격 : 25.705


- 학습이 안된 테스트 데이터라 정확한 예측값이 어려움 
- 실제가격 : 43.500, 예상가격 : 34.083 비싼 축인데 우리모델을 그렇게 까지 비싸야할까라고 판단함